## 🏷 예외처리

</br>

### 예외 클래스 계층

- **checked exception**
대처 코드 없으면 컴파일 안됨
*⇒ 처리코드가  필요! (빨간 줄)*
- **unchecked exception**
대처 코드 없어도 컴파일은 진행 (컴파일 시에 예상할 수 없는 예외)
*⇒ 예외 처리 코드 없어도 컴파일은 가능*

![예외계층](https://user-images.githubusercontent.com/59721896/183273767-eb1e3937-1bee-4135-a48e-486ace9c967f.png)

(파란 박스: unchecked exception, 빨간 박스: checked exception)

(이미지 출처: [https://jeeneee.dev/java-live-study/week9-exception-handling/](https://jeeneee.dev/java-live-study/week9-exception-handling/))

</br>

*👉 RuntimeException의 특징 !*

많은 경우 RuntimeException은 ‘처리’ 보다는 디버깅의 대상이다.

</br>

### try ~ catch 구문의 흐름

1. **JVM이 해당 Exception 클래스의 객체 생성, throw!**
throw new XXException
2. **exception을 처리할 수 있는 catch 블록에서 받아서 처리**
3. **정상적으로 처리되면 try-catch 블록 벗어나서 다음 문장 실행**
try 블록에서 어떤 예외도 발생하지 않았을 경우에는 catch 무시!

</br>

*👉 try ~ catch 구문의 중요성*

예외 처리 측면에서도 중요하지만,

catch 부분을 통해서 **예외에 대한 로깅을 남길 수 있다는 점**도 중요!

예외는 꼭 로깅해주어야 하는데, 아래 메서드들을 통해서 편리하게 로깅 남기는 것이 가능하다.

| 메서드 | 동작 |
| :--: | :--: |
| public String getMessage() | 발생된 예외에 대한 구체적인 메시지 반환 |
| public Throwable getCause() | 예외의 원인이 되는 Throwable 객체 또는 null 반환 |
| public void printStackTrace()  | 예외가 발생된 메서드가 호출되기까지의 메서드 호출 스택 출력 디버깅의 수단으로 주로 사용 |


</br>

### 다중 예외 처리

하나의 catch 블록에 여러 개의 catch 블록 추가 가능

⇒ 예외 종류별로 catch 블록 구성!

</br>

*👉 다중 예외 처리 진행 시 주의할 점!*

JVM이 던진 예외는 catch 문장 찾을 때 다형성이 적용됨

따라서 **상위 타입 예외가 먼저 선언되는 경우 뒤의 catch 블록은 동작할 기회가 없다!!**

상속 관계가 없는 경우에는 문제 없지만,

**상속 관계에서는 작은 범위에서 큰 범위순으로 정의하는 것이 귯! (자식에서 조상 순) ✨**

</br>

*👉 다중 예외 처리를 이용한 checked exception 처리 시 주의할 점!*

예외 처리는 가능한 구체적으로 하는 것이 베스트!!

```java
public void main(String[] args) {
		try {
			Class.forName("abc.Def"); // ClassNotFoundException
			new FileInputStream("Hello.java"); // FileNotFoundException
			DriverManager.getConnection("Hello"); // SQLException
		} catch(ClassNotFoundException e) {
			System.out.printf("클래스를 찾을 수 없습니다.: %s\n", e.getMessage());
		} catch(FileNotFoundException e) {
			System.out.printf("파일을 찾을 수 없습니다.: %s\n", e.getMessage());
		} catch(SQLException e) {
			System.out.printf("DB 접속 실패: %s\n", e.getMessage());
		}
		System.out.println("시스템 정상 종료");
	}
```

위의 예제를 확인하면 발생 가능한 예외들을 종류별로 분류한 것 확인 가능

Exception 클래스를 이용하여 하나로 처리하는 것도 가능하지만, 예외 별 상황을 

처리하거나 로깅하기에 부적합할 수 있기 때문에 **가급적 예외 별로 처리하는 것이 권장됨**

</br>

그러나 반대로, 심각하지 않은 예외를 굳이 세분화하는 것도 낭비일 수 있음

**하나의 catch 구문에서 상속관계가 없는 여러 개의 exception을 처리하는 것도 좋은 방법!**

```java
public void main(String[] args) {
		try {
			Class.forName("abc.Def");
			new FileInputStream("Hello.java");
			DriverManager.getConnection("Hello");
		} catch(ClassNotFoundException | FileNotFoundException e) {
			System.out.printf("자원을 찾을 수 없습니다.: %s\n", e.getMessage());
		} catch(SQLException e) {
			System.out.printf("DB 접속 실패: %s\n", e.getMessage());
		}
		System.out.println("시스템 정상 종료");
	}
```

</br>

### try ~ catch ~ finally

finally는 예외 발생 여부와 관계 없이 무조건! 실행됨 (try 블록에 진입했다면)

try문 중간에 return문을 만나도 finally 실행 후 리턴이 진행됨

```java
try {
		// exception이 발생할 만한 코드 - System 자원 사용
} catch(Exception e) {
		// XXException 발생 시 처리 코드
} finally {
		// try 블록에서 접근했던 System 자원의 안전한 원상복구
}
```

</br>

### finally 블록의 필요성

DB나 파일처리를 할 때, 자원을 열었다가 다시 닫지 않으면 (생성한 시스템 자원 반납하지 않으면)

장래에 resource leak이 발생할 수 있기 때문에 close 처리가 중요하다.

*⇒ OS별로 열 수 있는 파일의 개수가 정해져있기 때문!!!*

```java
FileInputStream fileInput = null;
		try {
			fileInput = new FileInputStream("abc.txt");
			fileInput.read();
		} catch(IOException e) {
			e.printStackTrace();
		} finally {
			if(fileInput != null) {
				try {
					fileInput.close();
				} catch(IOException e) {
					e.printStackTrace();
				}
			}
		}
```

따라서 위의 예제와 같이 finally 블록을 이용하여 close를 해주면 좋은데,

**close 메서드 자체가 IOException을 유발할 수 있으므로, 그 예외를 또 처리해주어야 하며**

자연스럽게 finally 블록은 지저분해질 수밖에 없다.

</br>

### try-with-resources

JDK 1.7 이상의 리소스에서 자동으로 close 처리를 해주는 기능!

```java
try(FileInputStream fileInput = new FileInputStream("abx.txt")) {
		System.out.println("FileInputStream이 생성되었습니다.");
		fileInput.read();
} catch(IOException e) {
		System.out.println("파일 처리 실패");
```

*⇒ try 선언문에 선언된 객체들에 대해서 자동으로 close를 호출하며 finally 역할을 대신한다.*

**단, 해당 객체들이 AutoCloseable 인터페이스를 구현해야 한다.**

**또한 해당 객체는 try 블록에서 다시 할당될 수 없다.**

</br>

### throws

메서드에서 처리해야 할 하나 이상의 예외를 호출한 곳으로 전달하는 것! **(처리 위임)**

- 예외가 없어지는 것x, 단순히 전달o
- 예외 전달받은 메서드는 다시 예외 처리의 책임이 발생됨
- 처리하려는 예외의 조상 타입으로도 throws 가능

</br>

**👉 checked exception과 throws**

![checkedException과throws](https://user-images.githubusercontent.com/59721896/183273844-f649c562-3eac-4757-b3a2-da70205e9120.png)

위와 같이, 예외가 발생할 수 있는 부분의 실행에 대한 메서드를 따로 작성하고,

해당 메서드에서 발생 가능한 예외를 **throws 하여 catch에 예외를 전달해주는 방식!!**

*⇒ checked exception은 반드시 try ~ catch 또는 throws 필요 !!*

</br>

**👉 unchecked exception과 throws**

![uncheckedException과throws](https://user-images.githubusercontent.com/59721896/183273893-efe44b4b-bd74-472b-a150-1fc0a22024ba.png)

*runtime exception의 경우 throws하지 않아도 예외 전달!!! ✨* 

그러나 결국 try ~ catch로 처리해야 한다는 점은 동일하다.

</br>

### 로그 분석과 예외 추적

printStackTrace 메서드 사용 시 스택 정보를 조회할 수 있는데,

최초 호출 메서드에서부터 예외 발생 메서드까지의 스택 정보를 알려준다.

여기서 꼭 확인해야 할 정보들 !

- 어떤 예외인지 **(예외 종류)**
- 예외 객체의 메시지는 무엇인지 **(예외의 원인)**
- 어디서 발생했는지 **(디버깅 출발점)**
→ 직접 작성한 코드를 디버깅 대상으로 삼을 것
→ 참조하는 라이브러리(java.xx 등)는 과감히 패스

![로그](https://user-images.githubusercontent.com/59721896/183273904-6253e426-bfd2-4cd3-9c81-5d473c9e8431.png)

</br>

### throws의 목적와 API 활용

메서드 선언부에 throws 키워드와 함께 발생 가능한 예외를 명시함으로써

프로그래머가 예외 발생 가능성을 인지하고, 그 예외에 대처할 수 있도록 한다.

*⇒ 처리는 try ~ catch, 예외 던지는건 throws라고 생각!!*

자바 API에서 제공하는 많은 메서드들 또한 사전에 예외가 발생할 수 있음을 선언부에 명시한다.

```java
public FileInputStream(String name) throws FileNotFoundException // 이런식으로!
```

</br>

### 메서드 재정의와 throws

메서드 재정의 시 조상 클래스 메서드가 던지는 예외보다 부모 예외를 던질 수 없음 !

</br>

## 🏷️ 사용자 정의 예외

</br>

### 사용자 정의 예외

API에 정의된 exception 이외에 필요에 따라 사용자 정의 예외 클래스를 작성할 수 있다.

대부분 Exception 또는 RuntimeException 클래스를 상속받아 작성한다.

- **checked exception 활용**
명시적 예외 처리 또는 throws 활용
⇒ 코드는 복잡해지지만 처리 누락 오류 등 발생 가능성 줄어듬
- **runtime exception 활용**
묵시적 예외 처리 가능
⇒ 코드가 간결해지지만 예외 처리 누락 가능성 발생

</br>

### 사용자 정의 예외의 장점

- **객체의 활용**
필요한 추가 정보, 기능 활용 가능
- **코드의 재사용**
동일한 상황에서 예외 객체 재사용 가능
- **throws 메커니즘의 이용**
중간 호출 단계에서 return 불필요

</br>

### 사용자 정의 예외 작성

![사용자정의예외](https://user-images.githubusercontent.com/59721896/183273924-5ed2d2c5-bf6c-4f1e-84c6-2741ddc89e41.png)

</br> 

### 예외 처리 시 주의 사항 정리

1. 다중 예외 처리 시, 상위 타입 예외가 먼저 선언되지 않았는지
2. finally 블록을 적절히 사용했는지 (ex. close)
3. try-with-resources의 try에서 해당 객체가 재할당 되지는 않았는지
4. checked exception 처리 시 throws 처리 했는지
5. 메서드 재정의 시 부모보다 상위 타입의 예외를 던지지 않는지
